File: sol/solution/0001-0100/0029. Divide Two Integers/Solution.dart
class Solution {
  int divide(int dividend, int divisor) {
    const intMax = 2147483647;
    const intMin = -2147483648;
    if (dividend == intMin && divisor == -1) return intMax;
    if (dividend == intMin && divisor == 1) return intMin;
    int a = dividend.abs();
    int b = divisor.abs();
    int res = 0;
    while (a >= b) {
      int temp = b, multiple = 1;
      while ((temp << 1) <= a) {
        temp <<= 1;
        multiple <<= 1;
      }
      a -= temp;
      res += multiple;
    }
    return ((dividend > 0) ^ (divisor > 0)) ? -res : res;
  }
}

File: sol/solution/0001-0100/0029. Divide Two Integers/Solution.ex
defmodule Solution do
  @spec divide(integer, integer) :: integer
  def divide(dividend, divisor) do
    int_max = 2_147_483_647
    int_min = -2_147_483_648
    cond do
      dividend == int_min and divisor == -1 -> int_max
      dividend == int_min and divisor == 1 -> int_min
      true ->
        a = abs(dividend)
        b = abs(divisor)
        res = loop(a, b, 0)
        if (dividend > 0) != (divisor > 0), do: -res, else: res
    end
  end
  defp loop(a, b, res) when a < b, do: res
  defp loop(a, b, res) do
    {temp, multiple} = double_loop(b, a, b, 1)
    loop(a - temp, b, res + multiple)
  end
  defp double_loop(_b, a, temp, multiple) when temp * 2 > a, do: {temp, multiple}
  defp double_loop(b, a, temp, multiple), do: double_loop(b, a, temp * 2, multiple * 2)
end